nuxeo-features/nuxeo-platform-rendition/nuxeo-platform-rendition-core/src/main/java/org/nuxeo/ecm/platform/rendition/service/RenditionServiceImpl.java

```diff
@@ -384,4 +396,104 @@ protected RenditionDefinition getAvailableRenditionDefinition(DocumentModel doc,
+                    Serializable sourceVersionableId = res.get(RENDITION_SOURCE_VERSIONABLE_ID_PROPERTY);
+
+                    String key = sourceVersionableId != null ? sourceVersionableId.toString() : sourceId;
+                    List<String> renditionRefs = liveDocumentRefsToRenditionRefs.get(key);
+                    if (renditionRefs == null) {
+                        renditionRefs = new ArrayList<>();
+                        liveDocumentRefsToRenditionRefs.put(key, renditionRefs);
+                    }
+                    renditionRefs.add(renditionRef);
+                }
+            }
+            return liveDocumentRefsToRenditionRefs;
+        }
+
+        private void removeStoredRenditions(Map<String, List<String>> liveDocumentRefsToRenditionRefs, List<String> liveDocumentRefs) {
+            if (liveDocumentRefs.isEmpty()) {
+                // no more document to check
+                return;
+            }
+
+            // compute the batch of source ids to check for existence
+            List<String> batchSourceIds = new ArrayList<>();
+            int limit = BATCH_SIZE > liveDocumentRefs.size() ? liveDocumentRefs.size() : BATCH_SIZE;
+            for (int i = 0; i < limit; i++) {
+                if (i >= liveDocumentRefs.size()) {
+                    break;
+                }
+                batchSourceIds.add(liveDocumentRefs.remove(i));
+            }
+
+            // retrieve still existing documents
+            List<String> existingSourceIds = new ArrayList<>();
+            String query = NXQLQueryBuilder.getQuery("SELECT ecm:uuid FROM Document WHERE ecm:uuid IN ?",
+                    new Object[] { batchSourceIds }, true, true, null);
+            try (IterableQueryResult result = session.queryAndFetch(query, NXQL.NXQL)) {
+                for (Map<String, Serializable> res : result) {
+                    existingSourceIds.add(res.get("ecm:uuid").toString());
+                }
+
+            }
+            batchSourceIds.removeAll(existingSourceIds);
+
+            List<String> renditionRefsToDelete = batchSourceIds.stream().map(liveDocumentRefsToRenditionRefs::get).reduce(
+                    new ArrayList<>(), (allRefs, refs) -> {
+                        allRefs.addAll(refs);
+                        return allRefs;
+                    });
+
+            if (!renditionRefsToDelete.isEmpty()) {
+                session.removeDocuments(
+                        renditionRefsToDelete.stream().map(IdRef::new).collect(Collectors.toList()).toArray(
+                                new DocumentRef[renditionRefsToDelete.size()]));
+            }
+
+            if (TransactionHelper.isTransactionActiveOrMarkedRollback()) {
+                TransactionHelper.commitOrRollbackTransaction();
+                TransactionHelper.startTransaction();
+            }
+
+            // next batch
+            removeStoredRenditions(liveDocumentRefsToRenditionRefs, liveDocumentRefs);
```

Comment:
If this is indeed a recursive tail call then Java won't optimize it and you'll get stack growth for each batch. Rework as a loop.